# crispRdesignR
Software used to design guide RNA sequences for CRISPR/Cas9 genome editing

This software aims to provide all scientifically pertinent information when designing guide RNA sequences for Cas9 genome editing. When provided a target DNA sequence for editing, a genome to check for off-targets in, and a genome annotation file (.gtf) to provide addition information about off-target matches it will out put information for two separate data tables. The first table contains all information on the generated sgRNA themselves (sgRNA sequence, PAM, Direction, Start, End, GC content, Presence of Homopolymers, Self Complementarity, Effciency Score (Doench 2016), and Genomic Matches). The second table contains all information on the found off-target sequences (Original sgRNA Sequence, Chromosome, Start, End, Number of Mismatches, Direction, CFD Scores, Matched Sequence, Gene ID, Gene Name, Sequence Type, and Exon Number). Additionally, a user may provide their own DNA libraries to search for off targets in and use a genome annotation file of their preference.

crispRdesignR is tested in R 3.4.4.

## Installation and dependencies

Steps to install crispRdesignR (tested in R version x...):

##### dependencies gbm, vtreat and stringr:

`install.packages("gbm")`

`install.packages("vtreat")`

`install.packages("stringr", repos='http://cran.us.r-project.org')`

##### dependencies BioStrings and BSgenome packages through Bioconductor:

`source("https://bioconductor.org/biocLite.R")`

`biocLite("Biostrings")`

`biocLite("BSgenome")`

`biocLite("AnnotationHub")`

##### Your genome(s) of interest from BSgenome. The example dataset uses the yeast genome

`biocLite("BSgenome.Scerevisiae.UCSC.sacCer2")`

##### Install crispRdesignR, where "path_to_directory" is the path of the decompressed crispRdesignR-master folder

`install.packages(path_to_directory, repos = NULL, type="source")`

## Usage

All of the required files must present in the working directory, this includes Rule_Set_2_Model.rds, CFD_Scoring.csv, and a user supplied genome annotation file (.gtf). Genome annotation files can be found here: https://useast.ensembl.org/info/data/ftp/index.html

##### sgRNA Design

All data is generated with a single function: `sgRNA_design(userseq, genomename, gtfname, userPAM, calloffs = TRUE, annotateoffs = TRUE)`

`userseq`: The target sequence to generate sgRNA guides for. Can either be a character sequence containing DNA bases or the name of a fasta/text file in the working directory.

`genomename`: The name of a geneome (in BSgenome format) to check for off-targets in. These genomes can be downloaded through BSgenome or compiled by the user.

`gtfname`: The name of a genome annotation file (.gtf) in the working directory to check off-target sequences against.

`userPAM`: An optional argument used to set a custom PAM for the sgRNA. If not set, the function will default to the "NGG" PAM. Warning: Doench efficieny scores are only accurate for the "NGG" PAM.

`calloffs`: If TRUE, the function will search for off-targets in the genome chosen specified by the genomename argument. If FALSE, off-target calling will be skipped.

`annotateoffs`: If TRUE, the function will provide annotations for the off-targets called using the genome annotation file specified by the gtfname argument. If FALSE, off-target annotation will be skipped.

##### sgRNA Data Retrieval

The data on the generated sgRNA sequences can be retrieved with: `getsgRNAdata(x)`

`x`: The raw data generated by `sgRNA_design()`

##### Off-Target Data Retrieval

The additional off-target data can be retrieved with `getofftargetdata(x)`

`x`: The raw data generated by `sgRNA_design()`

##### sgRNA Design Doench 2014 (outdated)

An older version of the sgRNA_design function: `sgRNA_design_Doench2014(userseq, genomename, gtfname, userPAM, calloffs = TRUE, annotateoffs = TRUE)`

Uses same arguments as `sgRNA_design`

###### Example:
`alldata <- sgRNA_design("GGCAGAGCTTCGTATGTCGGCGATTCATCTCAAGTAGAAGATCCTGGTGCAGTAGG", BSgenome.Scerevisiae.UCSC.sacCer2, "Saccharomyces_cerevisiae.R64-1-1.92.gtf.gz")`

`sgRNAdata <- getsgRNAdata(alldata)`

`offtargetdata <- getofftargetdata(alldata)`

###### Example:
`exampledata <- sgRNA_design("DAK1.fasta", BSgenome.Scerevisiae.UCSC.sacCer2, "Saccharomyces_cerevisiae.R64-1-1.92.gtf.gz", "NAG", calloffs = TRUE, annotateoffs = FALSE)`
